benchmarks/QueryBench.php

```diff
@@ -0,0 +1,103 @@
+use PhpBench\Benchmark\Metadata\Annotations\BeforeClassMethods;
+use PhpBench\Benchmark\Metadata\Annotations\BeforeMethods;
+use PhpBench\Benchmark\Metadata\Annotations\AfterMethods;
+use PhpBench\Benchmark\Metadata\Annotations\Iterations;
+use PhpBench\Benchmark\Metadata\Annotations\Revs;
+use PhpBench\Benchmark\Metadata\Annotations\Warmup;
+use PhpBench\Benchmark\Metadata\Annotations\OutputTimeUnit;
+use function Amp\Promise\wait;
+use function Amp\Promise\all;
+use function Amp\call;
+
+/**
+ * @BeforeMethods({"init"})
+ * @AfterMethods({"onAfterMethods"})
+ * @Iterations(1)
+ * @Revs(10)
+ * @Warmup(1)
+ * @OutputTimeUnit("milliseconds", precision=5)
+ */
+class QueryBench extends AbstractBench
+{
+    /** @var  ConnectionPool */
+    protected $connectionPool;
+    /** @var  Connection */
+    protected $connection;
+    /** @var  \PDO */
+    protected $pdoConnection;
+    protected $maxQueries = 10;
+    protected $poolLimit = 10;
+
+    public function init()
+    {
+        $config = ConnectionConfig::parseConnectionString("host=$this->host;user=$this->user;pass=$this->pass");
+        $this->connectionPool = new ConnectionPool($config, $this->poolLimit);
+        $connectionPromise = Connection::connect($config);
+        $this->connection = wait($connectionPromise);
+        $this->pdoConnection = new \PDO("mysql:host=$this->host;", $this->user, $this->pass);
+    }
+
+    public function onAfterMethods()
+    {
+        $this->connectionPool->close();
+        $this->connection->close();
+    }
+
+    public function benchPdoQueries()
+    {
+        foreach (range(1, $this->maxQueries) as $ii) {
+            $resultSet = $this->pdoConnection->query("SELECT $ii");
+            $resultSet->fetchAll();
+        }
+    }
+
+    public function benchSyncQueries()
+    {
+        wait(call(function() {
+            $connection = $this->connection;
+            foreach (range(1, $this->maxQueries) as $i) {
+                /** @var ResultSet $resultSet */
+                $resultSet = yield $connection->query("SELECT $i");
+                yield $resultSet->advance();
+            }
+        }));
+    }
+
+    public function benchAsyncQueries()
+    {
+        wait(call(function() {
+            $connection = $this->connection;
+            /** @var ResultSet[] $resultSets */
+            $resultSets = yield all(array_map(function ($i) use ($connection) {
+                return $connection->query("SELECT $i");
+            }, range(1, $this->maxQueries)));
+            yield all(array_map(function ($resultSet) {
+                /** @var ResultSet $resultSet */
+                return $resultSet->advance();
+            }, $resultSets));
+        }));
+    }
+
+    public function benchAsyncQueriesUsingPool()
+    {
+        wait(call(function() {
+            $connection = $this->connectionPool;
+            /** @var ResultSet[] $resultSets */
+            $resultSets = yield all(array_map(function ($i) use ($connection) {
```

Comment:
@trowski already I have faced with this problem. To avoid such situation I have added `@Warmup(1)` to create all needed connection during the first run and avoid connection overhead to be included in overall benchmark time.

Reply:
When I copy the body of `benchAsyncQueriesUsingPool` to `init`, it reduces the time for the benchmark relative to the others significantly. Can you explain what's going on there?

Reply:
@trowski in nutshell PHPBench does not count the time that was spent during `init` execution. 

To understand how it works need to describe how works Revs(revolutions), Iterations and BeforeMethods(read more http://phpbench.readthedocs.io/en/latest/writing-benchmarks.html)

During one Warmup(it is equivalent to one revolution) we create all needed connections in the pool(read more http://phpbench.readthedocs.io/en/latest/writing-benchmarks.html?highlight=Warmup)